File: Python/routes/albums.py
from flask import Blueprint, jsonify, request
import pymysql
import os
import boto3
from dotenv import load_dotenv
from utils.db import get_db_connection

# Cargar las variables de entorno
load_dotenv()

albums_bp = Blueprint('albums', __name__)

# Configurar AWS S3
s3_client = boto3.client(
    's3',
    region_name=os.getenv('AWS_REGION'),
    aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID'),
    aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY')
)

@albums_bp.route('/', methods=['GET'])
def get_albums():
    try:
        # Conectar a la base de datos
        connection = get_db_connection()
        with connection.cursor(pymysql.cursors.DictCursor) as cursor:
            cursor.execute('SELECT * FROM ALBUM')
            rows = cursor.fetchall()

            albums = [
                {
                    'id':           album['ID'],
                    'nombre':       album['NOMBRE'],
                    'usuario':      album['USUARIO'],
                    'creacion':     album['CREACION']
                }
                for album in rows
            ]
        connection.close()
        return jsonify(albums)
    except Exception as e:
        print(f"Error: {str(e)}")
        return jsonify({'error': str(e), 'message': 'Error al obtener los albumes'}), 500

@albums_bp.route('/<string:userId>', methods=['GET'])
def get_user_albums(userId):
    try:
        connection = get_db_connection()
        with connection.cursor(pymysql.cursor.DictCursor) as cursor:
            cursor.execute('SELECT * FROM ALBUM WHERE USUARIO = %s', (userId,))
            rows = cursor.fetchall()

            albums = [
                {
                    'id':           album['ID'],
                    'nombre':       album['NOMBRE'],
                    'usuario':      album['USUARIO'],
                    'creacion':     album['CREACION']
                }
                for album in rows
            ]
        connection.close()
        return jsonify(albums)
    except Exception as e:
        print(f"Error: {str(e)}")
        return jsonify({'error': str(e), 'message': 'Error al obtener los álbumes'}), 500

# Create album
@albums_bp.route('/crear', methods=['POST'])
def create_album():
    try:
        data = request.get_json()
        usuario = data.get('usuario')
        nombre_album = data.get('nombreAlbum')

        # Conectar a la base de datos
        connection = get_db_connection()
        with connection.cursor(pymysql.cursors.DictCursor) as cursor:
            cursor.execute('SELECT COUNT(*) AS count FROM ALBUM WHERE USUARIO = %s AND NOMBRE = %s', (usuario, nombre_album))
            count = cursor.fetchone()['count']
            if count > 0:
                return jsonify({'message': 'El nombre del álbum ya está en uso.'}), 400

            # Insertar álbum en la base de datos
            cursor.execute('INSERT INTO ALBUM (NOMBRE, USUARIO) VALUES (%s, %s)', (nombre_album, usuario))
            connection.commit()

        connection.close()

        return jsonify({'message': 'Álbum creado', 'id': cursor.lastrowid, 'nombre': nombre_album, 'usuario': usuario})

    except Exception as e:
        print(f"Error: {str(e)}")
        return jsonify({'error': str(e), 'message': 'Error al crear el álbum'}), 500

# Update album
@albums_bp.route('/actualizar', methods=['POST'])
def update():
    try:
        data = request.get_json()
        usuario = data.get('usuario')
        album = data.get('album')
        nombre = data.get('nombre')

        # Conectar a la base de datos
        connection = get_db_connection()
        with connection.cursor(pymysql.cursors.DictCursor) as cursor:
            # Insertar álbum en la base de datos
            cursor.execute('UPDATE ALBUM SET NOMBRE = %s WHERE USUARIO = %s AND ID = %s', (nombre, usuario, album))
            connection.commit()

        connection.close()

        return jsonify({'message': 'Álbum actualizado'})

    except Exception as e:
        print(f"Error: {str(e)}")
        return jsonify({'error': str(e), 'message': 'Error al actualizar el álbum'}), 500

# Delete album
@albums_bp.route('/eliminar', methods=['POST'])
def delete_album():
    try:
        data = request.get_json()
        usuario = data.get('usuario')
        album = data.get('album')

        # Conectar a la base de datos
        connection = get_db_connection()
        with connection.cursor(pymysql.cursors.DictCursor) as cursor:
            # Obtener imágenes del álbum antes de eliminar
            cursor.execute('SELECT IMAGEN FROM IMAGEN WHERE ALBUM = %s', (album,))
            images = cursor.fetchall()

            # Eliminar el álbum de la base de datos
            cursor.execute('DELETE FROM ALBUM WHERE USUARIO = %s AND ID = %s', (usuario, album))
            connection.commit()

        # Eliminar imágenes de S3
        if images:
            for img in images:
                image_url = img['IMAGEN']
                image_name = image_url.split('/').pop()
                # Eliminar objeto de S3
                s3_client.delete_object(
                    Bucket=os.getenv('AWS_BUCKET_NAME'),  # Usar variable de entorno para el nombre del bucket
                    Key=f'Fotos_Publicadas/{image_name}'
                )

        connection.close()

        return jsonify({'message': 'Álbum e imágenes eliminadas exitosamente'})

    except Exception as e:
        print(f"Error: {str(e)}")
        return jsonify({'error': str(e), 'message': 'Error al eliminar el álbum y sus imágenes'}), 500

File: Node/routes/albums.js
const { S3Client, DeleteObjectCommand } = require("@aws-sdk/client-s3");
const { Upload } = require("@aws-sdk/lib-storage");
const bcrypt = require('bcrypt');
const saltRounds = 10;

const express = require('express');
const router = express.Router();
const db = require('../utils/db');

const s3 = new S3Client({
    region: process.env.AWS_REGION,
    credentials: {
        accessKeyId: process.env.AWS_ACCESS_KEY_ID,
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
    }
});

router.get('/', async (req, res) => {
    try {
        const [rows] = await db.query('SELECT * FROM ALBUM;');
        const albums = rows.map (album => {
            return {
                id:             album.ID,
                nombre:         album.NOMBRE,
                usuario:        album.USUARIO,
                creacion:       album.CREACION
            }
        });
        res.json(albums);
    } catch (error) {
        res.status(500).json({ error: error.message, message: 'Error al obtener las imagenes' });
    }
    console.log('GET /imagenes');
});

router.get('/:userId', async (req, res) => {
    const userId = req.params.userId;
    try {
        const [rows] = await db.query('SELECT * FROM ALBUM WHERE USUARIO = ?;', [userId]);
        const albums = rows.map(album => {
            return {
                id: album.ID,
                nombre: album.NOMBRE,
                usuario: album.USUARIO,
                creacion: album.CREACION
            };
        });
        res.json(albums);
    } catch (error) {
        res.status(500).json({ error: error.message, message: 'Error al obtener los álbumes' });
    }
    console.log(`GET /albums/${userId}`);
});

router.post('/crear', async (req, res) => {
    try {
        const { usuario, nombreAlbum } = req.body;
        let [rows] = await db.query('SELECT COUNT(*) AS count FROM ALBUM WHERE USUARIO = ? AND NOMBRE = ?;', [usuario, nombreAlbum]);
        if (rows[0].count > 0) {
            return res.status(400).json({ message: 'El nombre del álbum ya está en uso.' });
        }
        const [result] = await db.query('INSERT INTO ALBUM (NOMBRE, USUARIO) VALUES (?, ?);', [nombreAlbum, usuario]);
        res.json({ message: 'Álbum creado', id: result.insertId, nombre: nombreAlbum, usuario: usuario });
    } catch (error) {
        res.status(500).json({ error: error.message, message: 'Error al crear el álbum' });
    }
    console.log('POST /album/crear');
});

router.post('/actualizar', async (req, res) => {
    try {
        const { usuario, album,  nombre} = req.body;

        // Insertar album en la base de datos
        const [result] = await db.query('UPDATE ALBUM SET NOMBRE = ? WHERE USUARIO = ? AND ID = ?;', [nombre, usuario, album]);
        res.json({ message: 'Álbum actualizado' });
    }
    catch (error) {
        res.status(500).json({ error: error.message, message: 'Error al actualizar el álbum' });
    }
    console.log('POST /album/actualizar');
});


router.post('/eliminar', async (req, res) => {
    const connection = await db.getConnection();  // Obtener una conexión de la pool de conexiones
    try {
        const { usuario, album } = req.body;

        // Iniciar la transacción
        await connection.beginTransaction();

        // Obtener las imágenes asociadas al álbum
        const [images] = await connection.query('SELECT IMAGEN FROM IMAGEN WHERE ALBUM = ?', [album]);
        if (images.length > 0) {
            for (const img of images) {
                const imageUrl = img.IMAGEN;
                const imageName = imageUrl.split('/').pop();
                const deleteParams = {
                    Bucket: process.env.AWS_BUCKET_NAME,
                    Key: `Fotos_Publicadas/${imageName}`,
                };

                // Eliminar imagen del bucket S3
                await s3.send(new DeleteObjectCommand(deleteParams));
            }

            // Eliminar imágenes de la base de datos
            await connection.query('DELETE FROM IMAGEN WHERE ALBUM = ?', [album]);
        }

        // Eliminar el álbum de la base de datos
        await connection.query('DELETE FROM ALBUM WHERE USUARIO = ? AND ID = ?', [usuario, album]);

        // Confirmar transacción
        await connection.commit();

        res.json({ message: 'Álbum e imágenes eliminadas exitosamente' });
    } catch (error) {
        // Revertir la transacción en caso de error
        await connection.rollback();
        console.error('Error al eliminar el álbum y sus imágenes:', error);
        res.status(500).json({ error: error.message, message: 'Error al eliminar el álbum y sus imágenes' });
    } finally {
        connection.release();  // Liberar la conexión de vuelta a la pool
    }
    console.log('POST /album/eliminar');
});

module.exports = router;